rpc: deprecate getaddressinfo label

jonatack opened the pull request:
This PR builds on #17578 (now merged) and deprecates the rpc getaddressinfo `label` field. The deprecated behavior can be re-enabled by starting bitcoind with `-deprecatedrpc=label`.

See http://www.erisian.com.au/bitcoin-core-dev/log-2019-11-22.html#l-622 and https://github.com/bitcoin/bitcoin/pull/17283#issuecomment-554458001 for more context.

Reviewers: This PR may be tested manually by building, then running bitcoind with and without the `-deprecatedrpc=label` flag while verifying the rpc getaddressinfo output and help text.

Next step: add support for multiple labels.

jonatack committed:
rpc: deprecate getaddressinfo label field

jonatack committed:
test: remove getaddressinfo label tests

jonatack committed:
test: getaddressinfo label deprecation test

laanwj commented:
I'm not sure about this. It wasn't too long ago that we deprecated "account" here. Now we're deprecating "label" too. For most purposes (like identifying invoices), assigning one label per address seems to be enough. If someone uses multiple labels per address then yes, a single "label" won't do. But I'm not sure it's worth the hassle for all current clients.

jonatack commented:
Thanks. It seems to me that the current state of both label and labels being returned might be confusing. Based on the previous two wallet meeting discussions, at the moment I imagine one of these options going forward:

1. deprecate `label`, and add multiple label capability to `labels` (seemed to be the most popular option)

2. keep `label`, and deprecate `labels`

3. keep `label`, and change `labels` to `tags` as a separate feature from the label

laanwj commented:
Maybe add 'label' only if there's a single label. Clients that don't use multiple-label-per-address functionality will be unaffected, while there's no need to guess what to return if there's multiple.

jnewbery commented:
Concept ACK. Returning the same information in two places in the same RPC call is redundant and confusing (especially when we add support for multiple labels per address, as I believe has always been the plan).

> change labels to tags as a separate feature from the label

Please no!

> Maybe add 'label' only if there's a single label.

I think it adds unnecessary complexity to clients to have the structure of the return object change based on values.

A few comments on the approach (same comments as for #17578):

- squash the first two commits (RPC changes and test changes) to avoid breaking git bisect
- add more text to the deprecated RPC help text (what the user should do to retain the old behaviour and when the old behaviour will be removed entirely)
- move the deprecation test to `rpc_deprecated.py`

laanwj commented:
> I think it adds unnecessary complexity to clients to have the structure of the return object change based on values.

Could return it as "" then. That already happens if there's no label so clients can cope with it.

I would be the last to complain about breaking backwards compatibility for good reason, but I really think we're going overboard here with RPC deprecation changes every release, but okay, I seem to be the only one with that opinion.

laanwj commented:
whoops :blush:  I think I'm confused here. joinmarket, for example, already switched to using `labels` in 2018
```
commit b52bc06acfead65c2b88ee4f64af44bc223656e8
Author: chris-belcher <chris-belcher@users.noreply.github.com>
Date:   Thu Sep 6 18:03:17 2018 +0100

    Switch over to using labels instead of accounts
    
    Bitcoin Core 0.17 deprecates the accounts feature and replaces it with
    labels. The RPC functions using accounts still work for use with older
    version of Core.
```
I thought it was something recently introduced, but it was added in 189e0ef33ec66f03abf85cfd4d0ede1a0c5c02d3 in the initial introduction of the labels API in 2016. I'm not exactly sure why a single "label" ever got added.

jonatack commented:
Thanks @jnewbery and @laanwj. I'm too recent here to have experienced the wrath of users but very much empathise and defer to more long-term contributors on whether 'tis better to suffer the arrows of causing breakage or leaving redundancy. Will update and rebase tomorrow.

jonatack committed:
doc: update release notes with getaddressinfo label deprecation

jonatack committed:
doc: address pr17578 review feedback

- https://github.com/bitcoin/bitcoin/pull/17578#discussion_r363975411
- https://github.com/bitcoin/bitcoin/pull/17578#discussion_r363969721
- https://github.com/bitcoin/bitcoin/pull/17578#discussion_r362703553

jonatack commented:
Rebased and addressed the remaining nits in #17578.

jnewbery reviewed (commented):
Code review ACK 77b340e77aa6faa03feedb04f18bb9a7f7e9e8a3

One minor doc nit, which you can ignore.

jonatack commented on doc/release-notes-17585.md:
```diff
@@ -0,0 +1,6 @@
+Deprecated or removed RPCs
```
Thanks! Good idea; done.

jonatack commented:
Thanks for reviewing and the info/suggestion, @jnewbery! Updated the last release note commit.

jnewbery commented:
ACK d3bc18408146e91b3836f72360ff6fa2420b6887

laanwj commented:
ACK d3bc18408146e91b3836f72360ff6fa2420b6887

meshcollider reviewed (commented):
utACK d3bc18408146e91b3836f72360ff6fa2420b6887
